Mathlib/NumberTheory/LSeries/Convolution.lean

```diff
@@ -0,0 +1,203 @@
+      rw [tsum_congr_set_coe h hS,
+        tsum_union_disjoint (Disjoint.set_prod_left disjoint_sdiff_right ..) ?_ ?_,
+        tsum_setProd_singleton_left 0 _ h, tsum_setProd_singleton_right _ 0 h]
+      · simp only [h, term_zero, zero_mul, tsum_zero, mul_zero, add_zero]
+      · simp only [h, Function.comp_def]
+        have : (fun x : {0} ×ˢ (@univ ℕ) ↦ term f s x.val.1 * term g s x.val.2) = 0 := by
+          ext p
+          rw [Set.mem_singleton_iff.mp p.prop.1, term_zero, zero_mul, Pi.zero_apply]
+        exact this ▸ summable_zero
+      · simp only [h, Function.comp_def]
+        have : (fun x : (@univ ℕ \ {0}) ×ˢ {0} ↦ term f s x.val.1 * term g s x.val.2) = 0 := by
+          ext p
+          rw [Set.mem_singleton_iff.mp p.prop.2, term_zero, mul_zero, Pi.zero_apply]
+        exact this ▸ summable_zero
+  -- now `n > 0`
+  have H : n.divisorsAntidiagonal = m ⁻¹' {n} := by
+    ext x
+    replace hn := hn.ne' -- for `tauto` below
+    simp only [Finset.mem_coe, mem_divisorsAntidiagonal, m, mem_preimage, mem_singleton_iff]
+    tauto
+  rw [← H, Finset.tsum_subtype' n.divisorsAntidiagonal h, term_of_ne_zero hn.ne',
+    convolution_def, Finset.sum_div]
+  refine Finset.sum_congr rfl fun p hp ↦ ?_
+  simp only [h]
+  obtain ⟨hp, hn₀⟩ := mem_divisorsAntidiagonal.mp hp
+  have ⟨hp₁, hp₂⟩ := mul_ne_zero_iff.mp <| hp.symm ▸ hn₀
+  rw [term_of_ne_zero hp₁ f s, term_of_ne_zero hp₂ g s, mul_comm_div, div_div,
+    ← mul_div_assoc, ← natCast_mul_natCast_cpow, ← Nat.cast_mul, mul_comm p.2, hp]
+
+end LSeries
+
+open Set in
+/-- The L-series of the convolution product `f ⍟ g` of two sequences `f` and `g`
+equals the product of their L-series, assuming both L-series converge. -/
+lemma LSeriesHasSum.convolution {f g : ℕ → ℂ} {s a b : ℂ} (hf : LSeriesHasSum f s a)
+    (hg : LSeriesHasSum g s b) :
+    LSeriesHasSum (f ⍟ g) s (a * b) := by
+  simp only [LSeriesHasSum]
+  have hsum := summable_mul_of_summable_norm hf.summable.norm hg.summable.norm
+  let m : ℕ × ℕ → ℕ := fun p ↦ p.1 * p.2
+  convert (HasSum.mul hf hg hsum).tsum_fiberwise m with n
+  exact term_convolution ..
+
+/-- The L-series of the convolution product `f ⍟ g` of two sequences `f` and `g`
+equals the product of their L-series, assuming both L-series converge. -/
+lemma LSeries_convolution {f g : ℕ → ℂ} {s : ℂ} (hf : LSeriesSummable f s)
+    (hg : LSeriesSummable g s) :
+    LSeries (f ⍟ g) s = LSeries f s * LSeries g s :=
+  (LSeriesHasSum.convolution hf.LSeriesHasSum hg.LSeriesHasSum).LSeries_eq
+
+/-- The L-series of the convolution product `f ⍟ g` of two sequences `f` and `g`
+equals the product of their L-series in their common half-plane of absolute convergence. -/
+lemma LSeries_convolution' {f g : ℕ → ℂ} {s : ℂ}
+    (hf : abscissaOfAbsConv f < s.re) (hg : abscissaOfAbsConv g < s.re) :
+    LSeries (f ⍟ g) s = LSeries f s * LSeries g s :=
+  LSeries_convolution (LSeriesSummable_of_abscissaOfAbsConv_lt_re hf)
+    (LSeriesSummable_of_abscissaOfAbsConv_lt_re hg)
```

Comment:
maybe `LSeries_convolution_eq_mul` is a better name?

Reply:
This is in line with `LSeries_add` (the L-series of a sum) and `LSeries_neg` (the L-series of a negative). But of course, here the opration on the right is not the same as on the left.
I have no strong feelings here, but would perhaps like further input.

Reply:
If you prefer this name that's fine